### PR TITLE
Implementing soft delete for storage bucket

### PR DESCRIPTION
**Pull request by 600lyy**
### Change description

Add support for soft delete policy for Google storage bucket.

The bucket's soft delete policy, which defines the period of time that soft-deleted objects will be retained, and cannot be permanently deleted. If it is not provided, by default Google Cloud Storage sets this to default soft delete policy`

This is the implementation for https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/2046

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

**Comment by yuwenma** [comment]
/lgtm

PR looks good. @600lyy could you fix the presubmit check?

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml [hunk 1]
```diff
@@ -287,6 +287,23 @@ spec:
                 required:
                 - retentionPeriod
                 type: object
+              softDeletePolicy:
+                description: The bucket's soft delete policy, which defines the period
+                  of time that soft-deleted objects will be retained, and cannot be
+                  permanently deleted. If it is not provided, by default Google Cloud
+                  Storage sets this to default soft delete policy.
+                properties:
+                  effectiveTime:
```
Comment (maqiuyujoyce) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml:296]: @yuwenma this field is an output-only field that gets inserted under `spec`. According to our previous discussion, we don't want to add any output-only spec fields any more - is my understanding correct?

File: pkg/test/resourcefixture/testdata/basic/storage/v1beta1/storagebucket/update.yaml [hunk 2]
```diff
@@ -23,3 +23,5 @@ spec:
   versioning:
     enabled: true
   lifecycleRule: []
+  softDeletePolicy:
+    retentionDurationSeconds: 0
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/storage/v1beta1/storagebucket/update.yaml:27]: I ran the test locally and found that `0` was not used in the update, and the value kept to be `604800`. Could you verify that this is a valid update; or whether setting the value to `0` is actually effective? AFAIK, we don't support setting a primitive-typed field to its empty value.

Comment (maqiuyujoyce) [reply pkg/test/resourcefixture/testdata/basic/storage/v1beta1/storagebucket/update.yaml:27]: I checked again and actually it worked.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml (hunk 1 above)
Comment (maqiuyujoyce) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml:296]: This is a workaround to hide output-only spec field but still support it under observed state: #2075 .

**Comment by justinsb** [comment]
Looks like I can do the rebase / cleanup; just pushed to the PR branch 🤞

**Comment by justinsb** [comment]
Looks like a (simple) rebase was all that was needed - thank you @600lyy 

/approve
/lgtm

@maqiuyujoyce I think we should merge this before #2075

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml (hunk 1 above)
Comment (yuwenma) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_storagebuckets.storage.cnrm.cloud.google.com.yaml:296]: Thank you @maqiuyujoyce for the quick workaround!

**Comment by maqiuyujoyce** [comment]
@justinsb as long as we don't cut a release, the order to merge #2051 and #2075 shouldn't matter.

/lgtm

**Pull request merged**